tests/MongoDB.Bson.Tests/Specifications/bson-corpus/BsonCorpusTestRunner.cs

```diff
@@ -0,0 +1,331 @@
+﻿/* Copyright 2020 MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using FluentAssertions;
+using MongoDB.Bson.IO;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+using Xunit;
+
+namespace MongoDB.Bson.Tests.Specifications.bson_corpus
+{
+    public class BsonCorpusTestRunner
+    {
+        [SkippableTheory]
+        [ClassData(typeof(TestCaseFactory))]
+        public void RunTestDefinition(JsonDrivenTestCase testCase)
+        {
+            var test = testCase.Test;
+            var testType = test["type"].AsString;
+
+            switch (testType)
+            {
+                case "valid": RunValidTest(test); break;
+                case "decodeErrors": RunDecodeErrorsTest(test); break;
+                case "parseErrors": RunParseErrorsTest(test); break;
+                default: throw new Exception($"Invalid test type: {testType}.");
+            }
+        }
+
+        // private methods
+        private void AsserException(Exception exception)
```

Comment:
Asser*t*Exception, sorry I had a typo in my suggestion

Reply:
I've seen it, but for some reason made it myself :)
Done.